src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs

```diff
@@ -0,0 +1,85 @@
+/// </summary>
+public static class OpenApiRouteHandlerBuilderExtensions
+{
+    /// <summary>
+    /// Adds an OpenAPI annotation to <see cref="Endpoint.Metadata" /> associated
+    /// with the current endpoint.
+    /// </summary>
+    /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
+    /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the endpoint.</returns>
+
+    public static RouteHandlerBuilder WithOpenApi(this RouteHandlerBuilder builder)
+    {
+        builder.Add(endpointBuilder =>
+        {
+            if (endpointBuilder is RouteEndpointBuilder routeEndpointBuilder)
+            {
+                var openApiOperation = GetOperationForEndpoint(routeEndpointBuilder);
+                if (openApiOperation != null)
+                {
+                    routeEndpointBuilder.Metadata.Add(openApiOperation);
+                }
+            };
+        });
+        return builder;
+
+    }
+
+    /// <summary>
+    /// Adds an OpenAPI annotation to <see cref="Endpoint.Metadata" /> associated
+    /// with the current endpoint and modifies it with the given <paramref name="configureOperation"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
+    /// <param name="configureOperation">An <see cref="Action"/> that mutates an OpenAPI annotation.</param>
+    /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the endpoint.</returns>
+    public static RouteHandlerBuilder WithOpenApi(this RouteHandlerBuilder builder, Func<OpenApiOperation, OpenApiOperation> configureOperation)
+    {
+        builder.Add(endpointBuilder =>
+        {
+            if (endpointBuilder is RouteEndpointBuilder routeEndpointBuilder)
+            {
+                var openApiOperation = GetOperationForEndpoint(routeEndpointBuilder);
+                if (openApiOperation != null)
+                {
+                    routeEndpointBuilder.Metadata.Add(configureOperation(openApiOperation));
+                }
+
+            };
+        });
+        return builder;
+    }
+
+    private static OpenApiOperation? GetOperationForEndpoint(RouteEndpointBuilder routeEndpointBuilder)
+    {
+        var pattern = routeEndpointBuilder.RoutePattern;
+        var metadata = new EndpointMetadataCollection(routeEndpointBuilder.Metadata);
+        var methodInfo = metadata.OfType<MethodInfo>().SingleOrDefault();
+        var serviceProvider = routeEndpointBuilder.ServiceProvider;
+
+        if (methodInfo == null || serviceProvider == null)
+        {
+            return null;
+        }
+
+        var hostEnvironment = serviceProvider.GetService<IHostEnvironment>();
+        var serviceProviderIsService = serviceProvider.GetService<IServiceProviderIsService>();
+        var generator = new OpenApiGenerator(hostEnvironment, serviceProviderIsService);
```

Comment:
Now that `OpenApiGenerator` is no longer a singleton that's resolved from DI does it need to be refactored to make better use of its `ParameterBindingMethodCache` instance, etc.? e.g. `RequestDelegateFactory` creates a static instance of `ParameterBindingMethodCache`. That said, given this is in a package it might not be feasible to share the instance without making the type public anyway.

Reply:
Yeah, the addition of this package certainly ups the case for https://github.com/dotnet/aspnetcore/issues/33644 and introduces the additional challenge of this particular codepath needing to return `ParameterLocation` instead of `BindingSource`.

At least for all the complex types, the new `IProvidesMetadata` stuff might help a bit but not sure what to do about primitives.